nvme_gw_server.py

```diff
@@ -528,74 +545,102 @@ def nvmf_subsystem_remove_host(self, request, context=None):
     def nvmf_subsystem_add_listener(self, request, context=None):
         """Adds a listener at the given TCP/IP address for the given subsystem."""
         self.logger.info({
-            f"Adding {request.trtype} listener at {request.traddr} : {request.trsvcid} for {request.nqn}"
+            f"Adding {request.gateway} {request.trtype} listener at {request.traddr} : {request.trsvcid} for {request.nqn}"
         })
 
-        try:
-            if request.trtype.lower() not in self.transports:
-                raise Exception(f"{request.trtype} transport is not enabled")
-
-            return_string = self.spdk_rpc.nvmf.nvmf_subsystem_add_listener(
-                self.spdk_rpc_client,
-                nqn=request.nqn,
-                trtype=request.trtype,
-                traddr=request.traddr,
-                trsvcid=request.trsvcid,
-                adrfam=request.adrfam,
-            )
-            self.logger.info(f"Status of add listener: {return_string}")
-        except Exception as ex:
-            self.logger.error(f"Add Listener failed: \n {ex}")
-            if context:
-                context.set_code(grpc.StatusCode.INTERNAL)
-                context.set_details(f"{ex}")
-            return pb2.req_status()
+        if not request.gateway or request.gateway == self.gateway_name:
+            try:
+                traddr = request.traddr
+                if not request.gateway:
+                    if request.trtype.lower() != "tcp":
```

Comment:
Why do none tcp transports require a gateway?

Reply:
I don't think `pcie` is allowed for NVMeoF. I see `fc`, `rdma`, `tcp` and `muser` in the code. At least for `rdma` it could be used in that way. For `fc` and `muser` we probably would need to allow to specify a `gateway_addr` that is not an ip. I don't have a strong opinion on it, on the other hand it would be nice if we could just support those out of the box. It seems like we would just need to remove trying to parse `gateway_addr` as an ip and let the create listener fail.